src/beast/python/beast/build/State.py

```diff
@@ -0,0 +1,96 @@
+# This file is part of Beast: http://github.com/vinniefalco/Beast
+#
+# This file copyright (c) 2015, Tom Ritchford <tom@swirly.com>
+# under the Boost software license http://www.boost.org/LICENSE_1_0.txt
+
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals)
+
+import collections, copy, os
+
+from . CheckForFastBuild import check_for_fast_build
+from . import Variant, Module, Target, Toolchain
+from .. System import SYSTEM
+
+
+class State(object):
+    """
+    Represents the state of a scons build environment.
+    """
+    def __init__(self, sconstruct_globals, environment, variant_tree, build_dir,
+                 environ=None):
+        class Sconstruct(object):
+            """Turn the SConstruct globals() back into a namespace,
+            so you can write things like self.sconstruct.COMMAND_LINE_TARGETS.
+            """
+            def __getattr__(self, key):
+                return sconstruct_globals[key]
+
+        self.sconstruct = Sconstruct()
+        self.env = self.sconstruct.Environment(**environment)
+        self.variant_tree = variant_tree
+        self.build_dir = build_dir
+        self.environ = environ or sconstruct_globals.get('ENV', os.environ)
+        self.aliases = collections.defaultdict(list)
+        self.msvc_configs = []
+        self.tags = [SYSTEM.platform.lower()]
+        if SYSTEM.linux:
+            self.tags.append('linux')
+
+        if SYSTEM.osx:
+            self.tags.append('osx')
+
+        self.subst = getattr(self.env, 'subst', lambda x: x)
+        check_for_fast_build(self)
+
+    def get_environment_variable(self, key, default=None):
+        return self.subst(self.environ.get(key, default))
+
+    def run(self, modules, targets):
+        targets = Target.Targets(*targets)
+        module = Module.compose(*modules)
+        module.before(self)
+
+        # Configure the toolchains, variants, default toolchain, and default
+        # target.
+        toolchains = Toolchain.detect(self)
+        if not toolchains:
+            raise ValueError('No toolchains detected!')
+
+        target_line = list(self.sconstruct.COMMAND_LINE_TARGETS)
+
+        for target, tags_list in targets.targets_to_tags(target_line).items():
+            if target.program_name:
+                for tags in tags_list:
+                    variant = Variant.Variant(
+                        self, tags, toolchains, target.program_name)
+                    variant.run(module)
+
+        for key, value in self.aliases.iteritems():
+            self.env.Alias(key, value)
+
+        module.after(self)
+
+
+DEFAULTS = {
+    'build_dir': 'build',
+
+    'environment': {
+        'toolpath': [os.path.join('src', 'beast', 'python', 'site_tools')],
```

Comment:
@rec: I need two changes to build my "hello world" project:

1) Change this to a relative path. What I had in mind:  https://github.com/seelabs/rippled/blob/53aaba32c4b32e10f545126ce25d073ecfffdc33/src/beast/python/beast/build/BuildState.py#L61-71 

2) Resolve the warning about "Two different environments were specified for target main.o"

(apologies if this is already on your to-do list. I left a comment on an outdated diff and I wanted to make sure this didn't get lost).